File: solutions/c/luhn/1/luhn.c
#include "luhn.h"

#include<string.h>

bool luhn(const char *num){
    if (!num)
        return false;
    int len=0;
    int sum=0;
    for(int i=strlen(num)-1; i>=0; i--){
        if(num[i]>='0' && num[i]<='9'){
            len++;
            char d = num[i]-'0';
            sum += (0==len%2) ? (d==9 ? 9 : 2*d % 9): d;            
        }
        else if(num[i]!=' ')
            return false;        
    }
    return len>1 ? (0==sum%10) : false;
}

File: solutions/python/luhn/1/luhn.py
class Luhn:
    def __init__(self, card_num):
        self.card_num = card_num

    def valid(self):
        len = 0
        sum = 0
        for ch in self.card_num[::-1]:
            if ch.isdigit():
                len += 1
                x = int(ch)
                if len%2==0:
                    if x==9:
                        sum += 9
                    else:
                        sum += (2*x)%9
                else:
                    sum += x                    
            elif ch!=' ':
                return False
        return len>1 and sum%10==0
            
